sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/TableHeaderReferenceCheck.java

```diff
@@ -0,0 +1,398 @@
+          Table.Header header = (Table.Header) column;
+          headers.get(col).add(header.id());
+        }
+      }
+    }
+    return headers;
+  }
+
+  private static List<Set<String>> findVerticalHeaders(Table table) {
+    List<Set<String>> headers = new ArrayList<>();
+    for (int i = 0; i < table.numberOfRows(); ++i) {
+      headers.add(new HashSet<>());
+    }
+    for (int row = 0; row < table.rows().size(); ++row) {
+      for (int col = 0; col < table.rows().get(row).size(); ++col) {
+        Table.Column column = table.rows().get(row).get(col);
+        if (column instanceof Table.Header) {
+          Table.Header header = (Table.Header) column;
+          headers.get(row).add(header.id());
+        }
+      }
+    }
+    return headers;
+  }
+
+  private static Map<TagNode, List<String>> findReferenceableHeadersPerColumnNode(Table table) {
+    List<Set<String>> horizontalHeaders = findHorizontalHeaders(table);
+    List<Set<String>> verticalHeaders = findVerticalHeaders(table);
+    Map<TagNode, List<String>> referenceable = new HashMap<>();
+    for (int row = 0; row < table.rows().size(); ++row) {
+      for (int col = 0; col < table.rows().get(row).size(); ++col) {
+        Table.Column column = table.rows().get(row).get(col);
+        if (column != null && hasHeaders(column.node())) {
+          List<String> headers = new ArrayList<>();
+          headers.addAll(horizontalHeaders.get(col));
+          headers.addAll(verticalHeaders.get(row));
+          referenceable.merge(column.node(), headers, (acc, val) -> { acc.addAll(val); return acc; });
+        }
+      }
+    }
+    return referenceable;
+  }
+
+  private static boolean isExistingHeader(Table table, String headerName) {
+    return table.rows().stream().flatMap(List::stream).filter(column -> column instanceof Table.Header)
+        .map(column -> (Table.Header) column).anyMatch(header -> headerName.equalsIgnoreCase(header.id()));
+  }
+
+  private static boolean isTable(TagNode node) {
+    return "TABLE".equalsIgnoreCase(node.getNodeName());
+  }
+
+  private static boolean isTableRow(TagNode node) {
+    return "TR".equalsIgnoreCase(node.getNodeName());
+  }
+
+  private static boolean isTableColumn(TagNode node) {
+    return "TD".equalsIgnoreCase(node.getNodeName());
+  }
+
+  private static boolean isTableHeader(TagNode node) {
+    return "TH".equalsIgnoreCase(node.getNodeName());
+  }
+
+  private static boolean hasHeaders(TagNode node) {
+    return node.hasProperty(HEADERS) && !isDynamicHeaders(node);
+  }
+
+  private static boolean isDynamicHeaders(TagNode node) {
+    return DYNAMIC_HEADERS.matcher(node.getPropertyValue(HEADERS)).find();
+  }
+
+  private static List<String> getHeaders(TagNode node) {
+    return Arrays.asList(node.getPropertyValue(HEADERS).split("\\s+")).stream()
+        .filter(header -> !header.isEmpty()).collect(Collectors.toList());
+  }
+
+  private static boolean hasRowSpan(TagNode node) {
```

Comment:
I find these methods redundant, they have no logic, just replacing instance method with static one, you can avoid using them by using `getRowSpan` and providing default value when attribute is not present